feat: add support for efficiently scheduling jobs from JS (#60)
graphile/worker: README.md

### Schedule a job:

There are two ways to schedule jobs:

1. From Node with the `publisher` api.
2. In the database directly with the `add_job` SQL function.

#### The Publisher API

Import `runPublisher` from `graphile_worker`, pass in your database configuration options (just like you would with `run` above.) and then pass a new job into `addJob` with a name and a payload!

```js
const { runPublisher } = require("graphile-worker");

async function main() {

>   const publisher = await runPublisher({ connectionString: "postgres:///" });

Reviewer: Lets make this more generic, so we can use it to expose more helpers in future. And "run" doesn't really feel like the right verb here.

```suggestion
  const utils = new Utils({ connectionString: "postgres:///" });
```

We should also make it clear (if necessary) that you should call `utils.end()` or similar when you're done with it.

I think we should also make an even more direct method for adding a job that wraps this:

```js
export async function addJob(config, ...args) {
  const utils = new Utils(config);
  try {
    return await utils.addJob(...args);
  } finally {
    utils.end();
  }
}
```

Using this directly a lot would be expensive due to the pool acquisition and releasing, but for the odd one-off might be better than passing a `utils` instance around?

Author: I see where you're coming from if your primary method of adding jobs is through the database, and the application only adds jobs now and then.

From my current architecture, and what I think the average JS server developer would expect, this `publisher` API would be my primary way of adding jobs.

```
    +-------------+ +-----------------+                   
    | Application | | Worker process  |                   
    +-------------+ +-----------------+                   
           |                 |                            
 Publisher |                 |                            
           |                 | Graphile CLI               
           |                 |                            
    +---------------------------------+                   
    |    Postgres                     |                   
    +---------------------------------+                   
               |     |                                    
               |     |                                    
               |     |                                    
               |------                                    
               Queries or triggers                        
```

In other words, I love the simplicity of the CLI, and don't actually intend to use the programmatic API for subscribing to jobs. (Not that the programmatic API isn't useful, but I think the CLI may be the way the majority of people will run the workers.)

Even though putting `addJob` into a `utils` scope would function just the same as what we have in this PR, I'd argue that even the name `utils` makes the approach of adding jobs through Node feel secondary. Do you think there's a compromise we can make about the way things are named, so that bootstrapping a publish-only connection for an application feels like a first-class pattern?

Please let me know if you think I'm missing the boat entirely here, as well 😅

Reviewer: I don't agree; users'd run something like:

```js
const workerUtils = new WorkerUtils({connectionString});

async function main() {
  // ... do stuff
  workerUtils.addJob(...);
  // ... do more stuff
  workerUtils.addJob(...);
  // ... yet more stuff
}
```

This `new Thing` approach is a common pattern in the Node ecosystem, it feels very natural, and it means you can create the utils in the global scope without having to `await` (and you can easily pass it around your app). It also means that you can ultimately `workerUtils.end()` if you need to clean things up (e.g. for tests).

As time goes on we can add more utils, like `workerUtils.getFailedJobs()`, `workerUtils.getQueueLength("QUEUE NAME HERE")` or `workerUtils.postponeJob(jobId, untilDate)` which I think makes a lot of sense for "utils" but wouldn't really make sense for "publisher".

The reason I'd go with `Utils` or `WorkerUtils` rather than `Worker` is because I don't want people to confusing instanciating it with actually running the worker. The reason I don't want to go with `Publisher` is because it's too specific, and I want to be able to add more helpers over time (as above).

If you can think of a better name that addresses these issues, I'm all ears.

Author: Okay, that makes sense to me. I’ll update the PR sometime soon here.

### Schedule a job:

There are two ways to schedule jobs:

1. From Node with the `WorkerUtils` or `addJob` APIs.
2. In the database directly with the `add_job` SQL function.

#### The WorkerUtils API

> Import `WorkerUtils` from `graphile_worker`, and make a new instance by passing in your database configuration options (just like you would with `run` above). What you get back is an instance of a class with a function called `addJob`, and a function called `end`.

Reviewer: ```suggestion
Import `WorkerUtils` from `graphile-worker`, and make a new instance by passing in your configuration:

- one of these to determine how to connect to the database:
  - `connectionString`: A PostgreSQL connection string to the database containing the job queue, or
  - `pgPool`: A `pg.Pool` instance to use
 
A `WorkerUtils` instance has the following methods:

- `end` - call this to release the `WorkerUtils` instance. It's typically best to use `WorkerUtils` as a singleton, so you often won't need this, but it's useful for tests or processes where you want Node to exit cleanly when it's done.
- `addJob(name: string, payload: JSON, options: TaskOptions)` - a method you can call to enqueue a job, it returns a promise that resolves to the job. `options` accepts the following parameters:
  - `run_at`?
```

- [ ] TODO: document options.

> The `addJob` function takes a name, and a payload, and returns a promise of a job that has been added to the queue. The `end` function closes the connection, don't forget to call it yourself if you're done adding jobs and don't want to keep the pool open anymore!

Reviewer: Now covered above.

```suggestion
```

```js
const { WorkerUtils } = require("graphile-worker");

async function main() {
  const workerUtils = new WorkerUtils({ connectionString: "postgres:///" });
  try {
    await workerUtils.addJob("calculate-life-meaning", { value: 42 });
  } finally {
    workerUtils.end();
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
```

> ##### The `addJob` standalone Javascript function

Reviewer: ```suggestion
##### The `addJob` standalone JavaScript function
```

> If you just want the fast and easy way to add a job, and you don't mind the cost of opening a DB connection pool, and then cleaning it up right away _for every job added_, there's the `addJob` convenience function:

Reviewer: ```suggestion
If you just want the fast and easy way to add a job, and you don't mind the cost of opening a DB connection pool, and then cleaning it up right away _for every job added_, there's the `addJob` convenience function. It takes the `WorkerUtils` configuration as the first argument, and then the remaining `WorkerUtils#addJob` arguments afterwards:
```

Reviewer: Note: I'm trying to be careful to use "configuration" for the WorkerUtils argument, and "options" for addJob to avoid confusion of just using "options" everywhere.
